Dynatrace log forwarder composition cloudwatch+kinesis+s3+lambda (#113)
awslabs/crossplane-on-eks: examples/upbound-aws-provider/composite-resources/serverless-examples/kinesis-lambda-s3-logs/README.md

# Example to deploy serverless architecture (kinesis log forwarder)
This example deploys the architecture depicted on the diagram. First, it applies the Crossplane XRD and Compositions. Then it applies the Claim to deploy the log forwarder with the AWS resources (Kinesis Data Firehose steam, S3 bucket, Lambda function), and then it applies a second Claim to subscribe the log forwarder to a CloudWatch log group. Last it will start sending logs to Lambda to be process which forwards the logs to a 3rd party log collector with a copy saved on S3.

![Serverless diagram](images/kinesis-lambda-s3.jpg)

## Pre-requisites:
 - [Upbound AWS Provider Crossplane Blueprint Examples](../../../README.md)

> Create a dynatrace account (there is a 16 day free trial) and create an API Key for log ingestion,

Reviewer: ```suggestion
- A Dynatrace account
Dynatrace provides a 16 day free trial. Once you are logged in, create an API Key for log ingestion,
```

for more information on creating the key and enabling log monitoring see the doc [Amazon CloudWatch Logs monitoring
](https://www.dynatrace.com/support/help/setup-and-configuration/setup-on-cloud-platforms/amazon-web-services/amazon-web-services-integrations/aws-service-metrics/cloudwatch-logs)

> You can the Dynatrace console in "Access Tokens" and generate a token or use the API to generate the token (sope log.ingest) with the followig command:

Reviewer: ```suggestion
Navigate to the Dynatrace console -> "Access Tokens" and generate a token or use the API to generate the token (scope log.ingest) with the following command:
```

```sh
curl -X POST "https://XXXXXXXX.live.dynatrace.com/api/v2/apiTokens" -H "accept: application/json; charset=utf-8" -H "Content-Type: application/json; charset=utf-8" -d "{\"name\":\"lambda-ingest-logs\",\"scopes\":[\"logs.ingest\"]}" -H "Authorization: Api-Token XXXXXXXX"
```

> Create an AWS bucket and upload the dynatrace lambda code, you can download latest version from github repository [github.com/dynatrace-oss](https://github.com/dynatrace-oss/dynatrace-aws-log-forwarder/releases)

Reviewer: ```suggestion
- S3 Bucket to store the lambda code
Create an AWS bucket and upload the Dynatrace lambda code, you can download latest version from github repository [github.com/dynatrace-oss](https://github.com/dynatrace-oss/dynatrace-aws-log-forwarder/releases)
```

lambda-invoke.iampolicy.awsblueprints.io                 22m
read-kms.iampolicy.awsblueprints.io                      22m
read-s3.iampolicy.awsblueprints.io                       22m
read-sqs.iampolicy.awsblueprints.io                      22m
s3.firehose.upbound.awsblueprints.io                     38h
s3.lambda.aws.upbound.awsblueprints.io                   22m
s3bucket.awsblueprints.io                                22m
subscriptionfilter.upbound.awsblueprints.io              22m
write-cloudwatch-metrics.iampolicy.awsblueprints.io      22m
write-firehose-s3.iampolicy.awsblueprints.io             22m
write-firehose.iampolicy.awsblueprints.io                22m
write-s3.iampolicy.awsblueprints.io                      22m
write-sqs.iampolicy.awsblueprints.io                     22m
```

## Update and apply Environment Config

Use the file template `environmentconfig-tmpl.yaml` to create a file `environmentconfig.yaml`

Set the variables `DYNATRACE_ENV_URL` and `DYNATRACE_API_KEY` in the following command with your valid values.

```sh
export DYNATRACE_ENV_URL="https://XXXXXXXX.live.dynatrace.com"
export DYNATRACE_API_KEY="dt0c01.XXXXXXXX"

> export S3_BUCKET="lambda-uploads-carrlos"

Reviewer: ```suggestion
export S3_BUCKET=<replace-me-with-s3-bucket-name>
```
